### ISSUE TITLE
Cross Compile - Static linking doesnt work

### ISSUE THREAD

**Issue by DJDan** [aircrack-ng/aircrack-ng#2040]
I am trying to cross compile from macOS Mojave 10.14.4 using "musl-cross-make" (Musl 1.1.21) to OpenWrt Mipsel (Musl 1.1.19) based router but when running the file on the router I get errors.

I tried to use --enable-static --disable-shared when cross compiling.
and export LDFLAGS=-static

When running on my router I get the following errors.
Error relocating ./airodump-ng: getentropy: symbol not found
Error relocating ./airodump-ng: explicit_bzero: symbol not found

**Comment by rofl0r** [comment]
the key point here is that a static compile was tried

> I tried to use --enable-static --disable-shared when cross compiling.
and export LDFLAGS=-static

but a dynamic binary generated. this used to work in the past. i suspect libtool is interfering.

**Comment by jbenden** [comment]
The `--enable-static` *only* controls libraries within the project. It has no effect on external libraries.

There are other flags for enabling static linking against some of the system libraries.

Also, why not use the OpenWrt package of aircrack-ng? I just updated it not long ago....

**Comment by rofl0r** [comment]
> Also, why not use the OpenWrt package of aircrack-ng?

the purpose of source is to be compiled...

> The --enable-static only controls libraries within the project. It has no effect on external libraries.

right. therefore additionally the environment variable `LDFLAGS` was set to `-static` to produce a static linked binary (with the prerequisite that all aircrack-internal libs were built static due to --enable-static).

and indeed, in my own testbuild, this happens:

```
/bin/bash ./libtool  --tag=CC   --mode=link powerpc-linux-musl-gcc -isystem /home/rofl/x-prefix/powerpc/include/ -L/home/rofl/x-prefix/powerpc/lib -pthread -Wall -O3 -std=gnu99 -fno-strict-aliasing -Wpointer-arith -Wstrict-overflow=2 -Wstrict-prototypes -fvisibility=hidden -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds    -static -o kstats src/kstats/kstats-kstats.o -lpthread -lssl -lcrypto -L/home/rofl/x-prefix/powerpc/lib -lz   -ldl -lm 
libtool: link: powerpc-linux-musl-ar cru .libs/libptw.a  lib/ptw/libptw_la-aircrack-ptw-lib.o
powerpc-linux-musl-ar: `u' modifier ignored since `D' is the default (see `U')
libtool: link: powerpc-linux-musl-ranlib .libs/libptw.a
libtool: link: powerpc-linux-musl-gcc -isystem /home/rofl/x-prefix/powerpc/include/ -pthread -Wall -O3 -std=gnu99 -fno-strict-aliasing -Wpointer-arith -Wstrict-overflow=2 -Wstrict-prototypes -fvisibility=hidden -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds -o kstats src/kstats/kstats-kstats.o  -L/home/rofl/x-prefix/powerpc/lib /lib/libssl.so -L=/lib /lib/libcrypto.so -lpthread -lz -ldl -lm -pthread
/lib/libssl.so: error adding symbols: File in wrong format
collect2: error: ld returned 1 exit status
make[2]: *** [kstats] Error 1
```
we can see that libtool not only removes -static from the linker line, it pulls -L=/lib out of its ass and replaces -lssl with a hardcoded path to /lib/libssl.so etc

libtool should not be used

**Comment by rofl0r** [comment]
now even if i add the switch --with-sysroot=/home/rofl/x-prefix/powerpc to configure, to make libtool properly behave in respect to the location where it looks for .la files, it still fails to honor -static and silently removes it:

```
 $ CC="powerpc-linux-musl-gcc -isystem /home/rofl/x-prefix/powerpc/include/ -L/home/rofl/x-prefix/powerpc/lib"  LDFLAGS=-static ./configure --enable-static --disable-shared --host=powerpc-linux-musl --with-sysroot=/home/rofl/x-prefix/powerpc
...
$ make V=1
...
/bin/bash ./libtool  --tag=CXX   --mode=link powerpc-linux-musl-g++ -L/home/rofl/x-prefix/powerpc/lib -isystem /home/rofl/x-prefix/powerpc/include  -Wall -O3 -fvisibility=hidden -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds    -static -o aircrack-ng src/aircrack-ng/aircrack_ng-aircrack-ng.o src/aircrack-ng/aircrack_ng-linecount.o src/aircrack-ng/aircrack_ng-session.o libaccrypto.la libaircrack.la libaircrack-ce-wep.la   libptw.la -lpthread -lssl -lcrypto -L/home/rofl/x-prefix/powerpc/lib -lz   -lssl -lcrypto  libaircrack-ce-wpa.la -ldl -lm 
libtool: link: powerpc-linux-musl-g++ -isystem /home/rofl/x-prefix/powerpc/include -Wall -O3 -fvisibility=hidden -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds -o aircrack-ng src/aircrack-ng/aircrack_ng-aircrack-ng.o src/aircrack-ng/aircrack_ng-linecount.o src/aircrack-ng/aircrack_ng-session.o  -L/home/rofl/x-prefix/powerpc/lib ./.libs/libaccrypto.a -L=/lib ./.libs/libaircrack.a ./.libs/libaircrack-ce-wep.a ./.libs/libptw.a ./.libs/libaircrack-ce-wpa.a /home/rofl/x-prefix/powerpc/lib/libpcre.so /seed/home/rofl/aircrack-ng/.libs/libaircrack-osdep.a -lnl-tiny /home/rofl/x-prefix/powerpc/lib/libssl.so /home/rofl/x-prefix/powerpc/lib/libcrypto.so -lpthread -lz -ldl -lm -pthread -Wl,-rpath -Wl,/home/rofl/x-prefix/powerpc/lib
```

**Comment by rofl0r** [comment]
even if i add --static to LDFLAGS, which libtool doesn't recognize (so it fails to remove it from the command line in its attempts to sabotage all static linking attempts) it still replace e.g. -lssl from the linker command line with an absolute path to libssl.so, even though a libssl.a exists to and there's absolutely no reason to replace -lssl (which would work as intended) with an absolute path to a file.

**Comment by jbenden** [comment]
These are all known limitation, with some work-arounds possible... It *is* possible to static link this, I already do for packaging...

See the other `./configure` flags to specify static ssl/crypto lib.

**Comment by rofl0r** [comment]
> See the other ./configure flags to specify static ssl/crypto lib.

well, -lssl was only an example. libtool does it to any -l item, pcre, libnl-tiny, pcap etc

> These are all known limitation

another way of expressing that libtool is broken. why not simply remove it ?

> It is possible to static link this, I already do for packaging...

how do **you** do it ?

**Comment by aircrack-ng** [comment]
Use the OpenWrt package as mentioned earlier. As mentioned, please look into the `./configure` file.

This last few comments look more like support and support is handled in the forums.

This is pretty much an edge case, so it's low priority (a workaround exists). We'll look into this issue later.

**Comment by DJDan** [comment]
As per @Mister-X- request here are the steps used to cross compile.

1) Install "musl-cross-make" - https://github.com/richfelker/musl-cross-make

sudo apt-get update
sudo apt install gcc make libisl-dev
sudo apt install build-essential autoconf automake libtool pkg-config libnl-3-dev libnl-genl-3-dev libssl-dev ethtool shtool rfkill zlib1g-dev libpcap-dev libsqlite3-dev libpcre3-dev libhwloc-dev libcmocka-dev hostapd wpasupplicant tcpdump screen iw
sudo apt install libc6-dev libc6-dev-i386

cd $HOME
git clone https://github.com/richfelker/musl-cross-make.git
cd $HOME/musl-cross-make
git clone https://github.com/sabotage-linux/kernel-headers.git
cd $HOME/musl-cross-make/kernel-headers
make ARCH=mips DESTDIR=$HOME/musl-cross-make_compiler/mipsel-linux-muslsf install

cd $HOME/musl-cross-make
echo OUTPUT="$HOME/musl-cross-make_compiler" > config.mak
echo TARGET="mipsel-linux-muslsf" >> config.mak
echo LINUX_VER="" >> config.mak
echo # Recommended options for faster/simpler build: >> config.mak
echo COMMON_CONFIG += CFLAGS="-g0 -O0" CXXFLAGS="-g0 -O0" LDFLAGS="-s" >> config.mak
echo COMMON_CONFIG += --disable-nls >> config.mak
echo GCC_CONFIG += --enable-languages=c,c++ >> config.mak
echo GCC_CONFIG += --disable-libquadmath --disable-decimal-float >> config.mak
echo GCC_CONFIG += --disable-multilib >> config.mak
make -j4
make install

2) Create "cross.sh" in $HOME
chmod +x $HOME/cross.sh

===

test -z "$1" && {
echo  "USAGE: $0 DIRECTORY_WITH_SOURCECODE OPTIONS_FOR_CONFIGURE"
exit 1
}

printf 'EXPORTING STUFF...\n\n'

export SDK="$HOME/musl-cross-make_compiler"
export STAGING_DIR="$SDK/mipsel-linux-muslsf"
export CFLAGS="-isystem $STAGING_DIR/include -D_GNU_SOURCE -include limits.h -static --static"
export CXXFLAGS="-static --static"
export LDFLAGS="-L$STAGING_DIR/lib -static"
export HOST="mipsel-linux-muslsf"
export PATH="$SDK/bin:$PATH"
export PKG_CONFIG_LIBDIR=$STAGING_DIR/lib/pkgconfig

cd "$1"
shift

test -e Makefile && { printf 'CLEANING UP!\n' ; make clean ; }
test -x autogen.sh && { printf 'AUTOGEN!\n' ; sh autogen.sh; }

if test -x configure ; then
MAKE_OPTIONS=
printf '\n=================================\n'
printf "NOW RUNNING CONFIGURE ... \n"
printf '=================================\n\n'
./configure --host="$HOST" --prefix="$STAGING_DIR" "$@" || { echo "FIX YOUR BUILD" ; exit 1 ; }
else
MAKE_OPTIONS="PREFIX=$STAGING_DIR prefix=$STAGING_DIR CC=$SDK/bin/${HOST}-gcc AR=$SDK/bin/${HOST}-ar RANLIB=$SDK/bin/${HOST}-ranlib"
fi

printf '=================================\n'
printf "NOW RUNNING MAKE (4 CORES)...\n"
printf '=================================\n\n'
make -j4 V=1 $MAKE_OPTIONS || {
	echo "OOPS, THAT WENT WRONG :("
	exit 1
}

printf '\n=================================\n'
printf "NOW RUNNING MAKE INSTALL...\n"
printf '=================================\n\n'
make install $MAKE_OPTIONS

3) Install aircrack-ng
cd $HOME
git clone https://github.com/aircrack-ng/aircrack-ng.git

4) Cross Compile dependencies (libnl-tiny & libressl)
cd $HOME
git clone https://github.com/sabotage-linux/libnl-tiny.git
wget https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-2.9.1.tar.gz
tar -xzf libressl-2.9.1.tar.gz

$HOME/cross.sh $HOME/libnl-tiny
$HOME/cross.sh $HOME/libressl-2.9.1

4) Cross Compile
$HOME/cross.sh $HOME/aircrack-ng --enable-static --disable-shared

NOTE:
The following must be applied before trying to cross compile aircrack-ng
https://github.com/aircrack-ng/aircrack-ng/pull/2039
=> build/m4/ax_check_openssl.m4 needs OPENSSL_LIBS="-lssl -lcrypto"

When running the cross compiled binaries on my router 
OpenWrt 18.06.2 r7676-cddd7b4c77) (v4.14.95) (rampis/mt7621) (Xiaomi Mi Router 3G)
(Musl 1.1.19) I get the following errors.

Error relocating ./airodump-ng: getentropy: symbol not found
Error relocating ./airodump-ng: explicit_bzero: symbol not found

NOTE: copying libc.so (musl-1.1.21) from the "musl-cross-make" directory manually into the router aircrack directory and running that before the airodump/aircrack etc... binaries works but I would like to know how to properly static link instead.

**Comment by rofl0r** [comment]
> echo TARGET="mipsel-linux-muslfs" >> config.mak

this should be muslsf i.e. **sf** as in softfloat, not fs

> --build="$BUILD"

you can remove the --build stuff and the BUILD env var. configure detects it automatically

> Error relocating ./airodump-ng: getentropy: symbol not found
> Error relocating ./airodump-ng: explicit_bzero: symbol not found

this is because your crosscompiler has musl 1.1.21 in it, whereas your router has musl 1.1.19 installed. because libtool sabotages a static build, you get a dynamic build and the result uses all symbols of musl 1.1.21' libc.so (those 2 are not yet in 1.1.19)

**Comment by rofl0r** [comment]
btw both @jbenden and @Mister-X- aka @aircrack-ng might be interested in my article about build systems, which explains cross-compile semantics: https://sabotage-linux.github.io/blog/5/

**Comment by aircrack-ng** [comment]
> this is because your crosscompiler has musl 1.1.21 in it, whereas your router has musl 1.1.19 installed. because libtool sabotages a static build, you get a dynamic build and the result uses all symbols of musl 1.1.21' libc.so (those 2 are not yet in 1.1.19)

So, there's no problem then?

**Comment by rofl0r** [comment]
https://github.com/aircrack-ng/aircrack-ng/issues/2040#issuecomment-483917927

again, the point here is that a static build was requested and a dynamic build produced. that the dynamic build doesn't work at runtime due to some other issue is irrelevant.

**Comment by jbenden** [comment]
Are you sure it didn't need `-static-libgcc -static-libstdc++` in the flags to get a static C & C++ library link?

**Comment by jbenden** [comment]
... also curious what the output of `ldd` is for the various binaries.

**Comment by DJDan** [comment]
root@OpenWrt:~# ldd aircrack-ng
	/lib/ld-musl-mipsel-sf.so.1 (0x77f25000)
	libnl-tiny.so => /usr/lib/libnl-tiny.so (0x77f0e000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x77db6000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x77d92000)
	libc.so => /lib/ld-musl-mipsel-sf.so.1 (0x77f25000)
Error relocating aircrack-ng: explicit_bzero: symbol not found
Error relocating aircrack-ng: getentropy: symbol not found

root@OpenWrt:~# ldd airodump-ng
	/lib/ld-musl-mipsel-sf.so.1 (0x77f2e000)
	libnl-tiny.so => /usr/lib/libnl-tiny.so (0x77f17000)
	libc.so => /lib/ld-musl-mipsel-sf.so.1 (0x77f2e000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x77ef3000)
Error relocating airodump-ng: getentropy: symbol not found
Error relocating airodump-ng: explicit_bzero: symbol not found

**Comment by rofl0r** [comment]
> Are you sure it didn't need -static-libgcc -static-libstdc++ in the flags to get a static C & C++ library link?

yeah. if -static would end up in the link command line, as it should, gcc would use libgcc_eh.a instead of libgcc_s.so automatically.

see the command line in my comment from https://github.com/aircrack-ng/aircrack-ng/issues/2040#issuecomment-484173040 : 
libtool turns a perfectly valid command like this:

`gcc foo.o bar.o -o baz -static -lssl -lcrypto`

into 

`gcc foo.o bar.o -o baz /usr/lib/libssl.so /usr/lib/libcrypto.so`.

there's no way this can ever produce a static binary. libtool should not remove `-static`, nor should it transform `-lssl` (or any other) into an absolute pathname to a library. these bugs are all known since a decade, but libtool is essentially unmaintained. the only way to fix it is to get rid of libtool or replace it with something that does the job properly. in my distro i use a libtool version i manually bugfixed. 
https://github.com/sabotage-linux/sabotage/blob/master/KEEP/libtool.cross
https://github.com/sabotage-linux/sabotage/blob/master/KEEP/libtool.cross.gen

btw my pr #2039 is a prerequisite to get this working. without merging that one first all attempts to reproduce this are futile.

**Comment by jbenden** [comment]
With PR #2039 applied, I am able to build static binaries.

The biggest key in it is to set `CFLAGS="-static --static" CXXFLAGS="-static --static"`. Also, there were small problems in finding archives; but nothing that one shouldn't be able to overcome.

**Comment by jbenden** [comment]
Merged #2039 just now...

**Comment by DJDan** [comment]
@jbenden Aircrack/Airodump now static links perfectly on OpenWrt. Thanks
I have also updated the instructions above with your extra parameters should anyone else wish to cross compile on Ubuntu.
